### ISSUE TITLE
[Proposal] Move actor timer and reminder functionality into Dapr's control plane

### ISSUE THREAD

**Issue by johnewart** [dapr/dapr#5403]
/area runtime

/area placement

### What does this propose?

This proposal suggests that we make some structural changes to the actor system in Dapr in order to improve performance and increase observability. In particular, this proposal focuses on the reminder and timer storage and dispatch model and the gRPC communication between daprd and the placement service as well as metrics to be added for improved observation, performance monitoring and diagnosis. The primary goal of this proposal is to move the code that is responsible for actor management into the control plane to simplify and centralize the actor subsystem.

#### Increase observability of the actor system 

This work requires the implementation of a metrics registry (see [[Metrics Registry]]), however, assuming that exists, the actor system will add new metrics to Dapr: 

Daprd: 
* Actor invocations (counters and timers) tagged with actor type and where they were invoked (possibly even ID?)
* Placement data interactions
	* Lock metrics (timers / counters)
	* Queries (hits and misses, timers)
	* Misses
	* Table sync operations
	* Table update operations (i.e actor type registration)

Placement service
* Table information
	* Number of actors types registered (gauge)
	* Number of current actors being placed (gauge)
	* Update metrics (timer)
	* Dissemination metrics (timers, counters)
* Reminder / timer information
	* Active reminders (gauges per time bucket)
	* Reminder dispatch performance (timers)
* Raft activity 
	* Synchronization metrics (timers)
	* Operation counts 
	* Health data

#### Move reminder storage and dispatch to placement service

Rather than have daprd manage reminder partitioning and storage, we should leverage the placement service's ability to do consistent hashing for placement / execution of reminders as well. This would allow for a more even distribution of reminders as well as flexible execution / scheduling of reminders if the number of sidecars grow and shrink. It also reduces the burden of `daprd` to manage reminder scheduling and will provide a single control mechanism for dispatching reminders to actor nodes. 

### Why do we need this?

As it stands there are performance-related issues that make it challenging to employ a large number of actor types or a large quantity of reminders. Currently, the time it takes to register new reminders increases with the number of reminders already in the system. 

Below we have a chart that shows a correlation between the number of reminders that have been registered and the CPU and memory resources required of `daprd`. 

![](https://raw.githubusercontent.com/johnewart/dapr-proposals/main/reminder-resources.png)

The right-most plot shows the total number of reminders that have been registered over time, while the left-most plot shows the % of CPU used by the `daprd` process and the middle plot shows the memory consumed by `daprd`. 

The process for testing the system was as follows:

1. Register N reminders, with a due-date and period such that they will fire twice to avoid registering reminders while executing them constantly, using the Go SDK 
2. Record memory and CPU usage by sampling the resources used by the `daprd` process in the process table
3. Allow some reasonable amount of time after creating the total number of reminders for collecting data
4. Programmatically remove the reminders that were created. 
5. Repeat with a larger value of N

This process was then repeated for increasing numbers of N, the values of which are the peaks of the line plot on the right (i.e 100, 500, 1K, 2K, 5K, 10K).

The end result is that creating more than a few hundred reminders quickly begins to consume large amounts of CPU time, as well as increases the amount of time it takes to register new reminders (as is shown by the increasing spacing along the X axis for each peak / trough on the right).

Additionally, testing shows that communication between `daprd` and the placement service begins to fail if an application registers more than 600 types of workers. 

As such, operating a large number of actors (in quantity) or registering large quantities of reminders proves to be problematic in the current implementation. 

#### Why now?

As a new feature in Dapr, we are going to be implementing a workflow engine that will rely heavily on actor subsytem to provide a significant amount of functionality that is required to implement workflows. Implementing a workflow engine requires that Dapr be capable of scheduling and dispatching potentially thousands of reminders concurrently, in addition each workflow will likely have many different types of actions that they take, each of which would be modeled as an actor, and as such it is entirely possible that a system having even only dozens of workflows could exceed the current limit of ~600 different actor types (imagine 30 different workflows with 20 different steps each, for example) quite quickly. 

In addition, a workflow engine will need to have a good level of visibility into its activity. Users (and / or operators) will need to be able to observe how many workflows are being executed, where they are running, what their state is, who is participating in the workflows, etc. Which will require being able to expose more information about the placement of actors, their throughput, and so on. 

Therefore, in order to implement a workflow engine that can scale with the needs of users, the actor subsystem will need to be able to handle the level of work that is required to make this happen. 

### How will this work?

(For simplicity, Reminders and Timers are treated as the same here because they are effectively identical from the perspective of the placement service)

#### Placement service changes

Placement service will now be responsible for three new pieces of functionality: Reminder scheduling (i.e knowing what reminders or timer need to fire when), reminder storage and reminder dispatching. 

##### Scheduling 

Using a time-base bucketing system (https://github.com/johnewart/go-timescheduler), the leader in the placement cluster will keep track of all the reminders that are scheduled across the `daprd` nodes that are connected to it. This has the advantage of being performant (time is divided into N + 1 buckets, the +1 being "everything past the last bucket") by decreasing the amount of data that is being evaluated, as well as easy to bulk load from storage with O(n) cost to rebuild the structure. 

Internally, each placement service will keep synchronized copies of the timer and reminder data in the same way it stores the actor placement data but only the leader will keep track of the scheduling data. Because the primary will be responsible for dispatching (since that is the node that `daprd` clients remain connected to), the scheduling data is only needed by the primary node. If / when a new node takes over from the existing leader, it can quickly reconstruct the schedule itself from its copy of the persisted data. 

##### Storage

The placement service will persist and synchronize the timer and reminder data in a similar fashion to the way it manages and stores placement data. Currently this is done using BoltDB but the actual storage implementation details are not critical as long as the performance characteristics are sufficient for our read-heavy workload. The data that will now be stored in addition to placement data on-disk will include the timer and reminder raw data (actor, schedule, etc.) that is required to reconstruct the scheduling data in the event of a failover or recovery.

> If we choose to put a limit on the size of reminder / timer data then we can choose to store the entire reminder table in memory alongside the scheduling data because we will know the memory requirements w.r.t the number of reminder and timers being stored ahead of time. I do not believe we currently impose any limits on the size of the data attached to a reminder

_Notes about write order:_ This should work as-is without having to worry about merging data between nodes because the existing implementation of  communication between `daprd` and the placement service cluster is such that `daprd` will only talk to a node that is the leader. As a result, if we opt for writing to stable storage on the leader first and then sending the data to the other nodes in the cluster for delayed writes to their stable storage, data-loss should not pose a serious concern.

##### Dispatching

Each placement service will maintain a dispatching thread that is responsible for:
* Evaluating the schedule for any timers / reminders that need to fire
* Determining which actor node will receive a given dispatch (using existing placement data or creating it if none exists) 
* Issuing gRPC commands to the target node for the dispatch of the event
* Handling rescheduling of the event in question for future execution
* Recording metrics regarding timing and counts

##### Reminder / timer registration and dispatch 

With this new structure, `daprd` would be responsible solely for execution of reminders and timers, similar to how it handles actor invocations, not for scheduling or storage / partitioning of reminder and timer data. In order for this to happen, the control channel used to communicate with placement service will need to have a new message type for invoking the reminder or timer and accompanying logic to handle those messages. 

###### Actor registration

Currently, `daprd` sends all of the known actor types at once to the placement service; this results in a loss of communication channels when the number of distinct actor types reaches a threshold around 600 or so. The net result is that (currently) the gRPC messages are too large and it tries to re-send them but continually fails. The change here will be to perform the actor registration in smaller chunks in order to avoid messages that are too large. 

#### Actor Registration

![placement-reminder](https://raw.githubusercontent.com/johnewart/dapr-proposals/main/placement-register.png)

#### Reminder Registration

![placement-scheduler](https://raw.githubusercontent.com/johnewart/dapr-proposals/main/placement-scheduler-1.png)

#### Reminder Execution

![reminder-execution](
![](https://raw.githubusercontent.com/johnewart/dapr-proposals/main/placement-scheduler-execute.png)

### Measurement of success

(These numbers are back-of-the napkin type numbers and very much open to feedback or discussion) 

The goal of this change is to improve the performance of actor dispatch and registration by moving it away from daprd and into placement service. In order to ensure that we have not only improved performance  but also kept existing functionality we will need to have tests that prove both of these things. 

The goal for reminder registration is to be able to register new reminders in consistent time (i.e adding new reminders should not increase in duration proportionally to the number of reminder already registered). A reasonable target should be registering a new reminder or timer should take <= 10msec consistently, regardless of the number of currently registered reminders or timers. 

Timer execution needs to happen in a timely fashion; reminders and timers should fire as close to their scheduled time as possible and there should be little drift as the number of registered timers or reminders grows. For any scheduled event, dispatch time (i.e the time at which the message fires from the placement service) should happen within a maximum of 50msec of its scheduled time with little change as the number of reminders increases. In addition, the time to dispatch the event should happen in under 10msec. 

In addition, the application should be able to expose many different types of actors without crashing the system. While it is perhaps unreasonable to expect that an application would export millions of actor types, it is reasonable to expect that it should be able to export at least several thousand without causing the system to fail. As a rough estimate, an application should be able to register at least 20,000 types of actors without causing an issue (but we should discover where the upper bound is, if there is one)

##### TL;DR
* Reminder / timer registration should target <= 10msec consistently, regardless of number of existing reminders or timers
* Timer dispatch should be within 50msec of scheduled time
* Dispatch message should be delivered to `daprd` in < 10msec over gRPC
* Applications should be capable of registering at least 20,000 different types of actors

#### Alpha

* Control-plane reminders and timers are *opt-in* through a configuration / flag
* Automatic migration of data will not happen, users opting in to this new feature will start with a blank slate
* The existing reminder / timer components will continue to operate as-is with the current performance profile / behavior
* (Open for discussion) Tools to manually migrate data from existing actor store to control plane?

#### Beta

* Control-plane implementation will become the default implementation and uses can *opt-out* through configuration
* Automatic, non-destructive, migration of data will be enabled (i.e moving data from the existing actor state store to the control plane)
* The current (i.e in-sidecar) implementation will be available (but again, only for those choosing to opt-out of the new implementation)

#### Stable

* Control-plane implementation will be the default behavior
* Previous, non-control-plane, implementation will not be available and will be removed

### What needs to be done

- [ ] Code changes
        - [ ] Add metrics (to measure current behavior and performance as a baseline)
	- [ ] Create new gRPC messages for placement service dispatch commands
	- [ ] Add timer / reminder registration to placement service
	- [ ] Add persistence / dissemination of reminder data to placement service cluster
	- [ ] Add scheduling and dispatch engine to placement service
	- [ ] Change daprd to forward registrations to placement cluster
	- [ ] Add migration code to move existing reminder / timers to placement service
	- [ ] Add dispatch handling to `daprd`
	- [ ] Change `daprd` to send registered actors to placement service in blocks instead of all-at-once
- [ ] Test changes
	- [ ] Additional performance tests for reminder registration and execution
	- [ ] E2E tests ensuring that placement service properly handles execution of reminders 
	- [ ] Double-activation guard tests 
- [ ] Documentation changes
	- [ ] Outline where reminders and timers are stored / how they are handled

**Comment by ItalyPaleAle** [comment]
I like this proposal but with one condition: that the placement service gets rid of Raft once for all. This would become even more critical if you need to give the placement service more responsibilities.

Context: currently placement can run in two modes of operation: non-HA (single node, usually data is persisted on disk or it won't survive a restart of the container) and HA (exactly 3 nodes-not more, not less-that use Raft to agree on the state, and don't _necessarily_ need to persist it to memory).

As you're thinking of giving a more active role to the placement service, its availability becomes even more important. You likely don't want a single instance of the service or there's no HA whatsoever, and a downtime in the placement service would cause a downtime in the application.

With Raft, we're severely limited in how we can get HA done "right" however:

1. With Raft, you can't have just 2 nodes, which would otherwise be sufficient in most cases to get HA. This is because a 2-node Raft cluster would halt entirely (losing the ability to elect a leader) when just 1 node is down, so it's not HA at all.
1. When using Raft, you must have exactly 3 replicas. In theory, even odd number > 1 would work (so also 5, 7, etc), but the more you have, the more bandwidth is required. Raft is incredibly chatty and with 5 nodes, you have a lot of network traffic in the cluster (and in many situations nowadays, network is more limited than people think it is).
1. Because you are forced to use exactly 3 nodes, it also means you cannot scale to more if you needed to, for example to support more reminders.

**Comment by johnewart** [comment]
I am in favor of this idea, but I was deliberately trying to avoid scope explosion for this proposal for the sake of timeliness. That being said, it would be feasible (and possibly even simpler in some ways) to change placement service (placement v2?) to operate differently for both placement data as well as reminders and timers (i.e without raft) and rather than a feature flag in `daprd`, placement service would just advertise itself differently and `daprd` could switch its behavior accordingly.

**Comment by yaron2** [comment]
Removing Raft is a separate discussion and should not be considered as a prerequisite for this proposal.

> and a downtime in the placement service would cause a downtime in the application.

This is a false statement, if the placement is down applications do not experience downtime, and to be more specific, applications invoking actors and running actor hosts do not experience downtime either.

I am willing to accept a PR that incorporates the changes proposed here irrespective of Raft changes.

**Comment by yaron2** [comment]
@johnewart what are your thoughts of moving this functionality to a separate, new service that has the sole responsibility of handling reminders and that can scale independently of placement?

**Comment by ItalyPaleAle** [comment]
> With the changes proposed here you may be getting reminder disruptions.

Yup, that's what I was referring to. Depending on the application, that could cause some degree of service degradation, if not a full outage.

> what are your thoughts of moving this functionality to a separate, new service that has the sole responsibility of handling reminders and that can scale independently of placement?

I think that makes sense. On K8s, this could use the K8s leader election APIs to elect a leader irregardless of the number of replicas (could even be an even number, hooray!). On self-hosted, could relay on the dapr-operator to perform leader election?

**Comment by johnewart** [comment]
As long as we are okay with adding more services to the control plane I am fine with adding a reminder service (I am in favor of smaller services as they are easier to work on anyhow). It will need to be able to place the actors onto nodes and thereby be dependent on placement service (not a problem, it can just retry for a bit and fall over if placement service isn't ready yet).

**Comment by ItalyPaleAle** [comment]
I personally think adding an extra service is not ideal (as it adds more operational complexities) UNLESS they can benefit from independent scaling, which seems to be the case here.

**Comment by olitomlinson** [comment]
Is there any update on this? I understand this a blocker for a preview of embedded Workflows being available in 1.10 however this is not triaged into tho 1.10 milestone?

**Comment by olitomlinson** [comment]
Hey @johnewart - is there any movement on this issue? Thanks!

**Comment by ScottArbeit** [comment]
Want. 🙂

**Comment by olitomlinson** [comment]
@yaron2 this can be closed now that distributed scheduler is available?

**Comment by yaron2** [comment]
> @yaron2 this can be closed now that distributed scheduler is available?

Indeed
